### ISSUE TITLE
SwitchEventStream breaks parent Signal's lazy-initial-value contract

### ISSUE THREAD

**Issue by pbuszka** [raquo/Laminar#103]
I found a very strange behavior in my system. For now I have not created a small repro but issue seems very serious so I give you a recap of my trying to narrow down the problem. 
```
    val openSessionRequest = new EventBus[ImportSourceEntry]

    val importFileProcessApiClient1 =
      appCanvas.$clients.value.map(_.map(_.importFileProcessApiClient))

    val importFileProcessApiClient2 =
      appCanvas.$clients.value.map(_.map(_.importFileProcessApiClient))

    val testResult = Val("test1")
      .combineWith(importFileProcessApiClient2)
      .flatMap { t =>
        EventStream.fromValue("test")
      }

```
... later in code I put some monitoring of what's going on
```
              div(
                importFileProcessApiClient --> { client =>
                  error(s"client monitoring mapped val 1 ${client}")
                },
                importFileProcessApiClient2 --> { client =>
                  error(s"client monitoring mapped val 2 ${client}")
                },
```
appCanvas.$clients is an SignalOptionT from laminext so a thin wrapper around Signal[Option[T]] and it's value starts with None and then later changes to Some(..)

so importFileProcessApiClient1 and importFileProcessApiClient2 are identical, the only difference is that importFileProcessApiClient2 is used in constructing testResult

what I got is
'client monitoring mapped val 1 Some(feecompass.app.panel.web.app.AppCanvas$$anon$18@37)'
'client monitoring mapped val 2 None'

if I remove the usage of importFileProcessApiClient2 i got what's expected
'client monitoring direct mapped 1 Some(feecompass.app.panel.web.app.AppCanvas$$anon$18@37) '
'client monitoring direct mapped 2 Some(feecompass.app.panel.web.app.AppCanvas$$anon$18@37) '

if I switch the usage of 1 and 2 the problem appears on 1
'client monitoring mapped val 1 None '
'client monitoring mapped val 2 Some(feecompass.app.panel.web.app.AppCanvas$$anon$18@35)'

Now the strange part testResult is *not* used anywhere. The trigger for the problem is
```
      .flatMap({ case (_, _) =>
        EventStream.fromValue("test")
      })
```
if I comment it out then it works as expected. 
If testResult is used for instance in `div(child <-- testResult.map(t => div(t.toString)), ...` the wrong behavior stays the same.

On the other hand switching to `FlattenStrategy.ConcurrentStreamStrategy` resolves the problem. 

It looks as if adding that flatMap with FlattenStrategy.SwitchStreamStrategy messes things up somehow and it's not related to actual usage of the resulting Observable but just the fact of constructing it. I will let you know if I find something more.

**Comment by yurique** [comment]
@pbuszka is this close enough to what you have?

```scala
    trait Client {
      def importFileProcessApiClient: EventStream[Resp]
    }
    trait Resp

    val $clients: Signal[Option[Client]] = ???

    val importFileProcessApiClient1 =
      $clients.map(_.map(_.importFileProcessApiClient))

    val importFileProcessApiClient2 =
      $clients.map(_.map(_.importFileProcessApiClient))

    val testResult = Val("test1")
      .combineWith(importFileProcessApiClient2)
      .flatMap { t =>
        EventStream.fromValue("test")
      }

    div(
      importFileProcessApiClient1 --> { client =>
        println(s"client monitoring mapped val 1 ${client}")
      },
      importFileProcessApiClient2 --> { client =>
        println(s"client monitoring mapped val 2 ${client}")
      }
    )
```

**Comment by pbuszka** [comment]
```
trait ImportFileProcessApiClient
case class Client(importFileProcessApiClient: ImportFileProcessApiClient))

// the rest is the same
```

**Comment by pbuszka** [comment]
any actual calls do not matter, just constructing that flatMap is enough to trigger the problem

**Comment by yurique** [comment]
I tried this, and I'm getting 
```
client monitoring mapped val 1 None
client monitoring mapped val 2 None
client monitoring mapped val 1 Some(...)
client monitoring mapped val 2 Some(...)
```

```scala
    class Client {
      def importFileProcessApiClient: ImportClient = new ImportClient
    }
    class ImportClient

    val clientsVar                       = Var(Option.empty[Client])
    val $clients: Signal[Option[Client]] = clientsVar.signal

    val importFileProcessApiClient1 =
      $clients.map(_.map(_.importFileProcessApiClient))

    val importFileProcessApiClient2 =
      $clients.map(_.map(_.importFileProcessApiClient))

    @unused val testResult = Val("test1")
      .combineWith(importFileProcessApiClient2)
      .flatMap { t =>
        EventStream.fromValue("test")
      }

    import scala.concurrent.duration._

      div(
        importFileProcessApiClient1 --> { client =>
          println(s"client monitoring mapped val 1 ${client}")
        },
        importFileProcessApiClient2 --> { client =>
          println(s"client monitoring mapped val 2 ${client}")
        }
      ),
      after(1.second) {
        clientsVar.set(Some(new Client))
      },
```

**Comment by pbuszka** [comment]
Maybe the fact that client is an effect of a previous chain of dependencies which is something like that:
```
val user = Var[Option[User]](None)
val entitlement = Var[Option[Entitlement]](None)
external user  --> user.set(Some(???))
external trigger  --> entitlement.set(Some(???))

val entitlementSignal = user.combineWith(entitlement)
...
val client = entitlementSignal.flatMap( ???: EventStream)

**Comment by yurique** [comment]
It shouldn't matter, at that point you have a `Signal` and that's it.

Btw, is that a typo in you example here?
```scala
                importFileProcessApiClient --> { client =>
                  error(s"client monitoring mapped val 1 ${client}")
                },
```
(`importFileProcessApiClient` without `1`)

**Comment by pbuszka** [comment]
It is quite complex leading to the point where I found the problem (user credentials, then getting entitlement based on crednetials, then geting dataToken obtained by an apiCall) and the final result is a Signal[Option[Client]] which contains all tokens and permission necessary

**Comment by pbuszka** [comment]
copy paste before I changed to 1 & 2 but to see the diference clearly.

**Comment by yurique** [comment]
> `val client = entitlementSignal.flatMap( ???: EventStream)`

Feels like it should be a `Signal` as well here.

To be honest, I've never tried doing `signal.flatMap(???: EventStream)` :)

**Comment by pbuszka** [comment]
I will try to narrow down which part upstream geenrates the problem.

**Comment by yurique** [comment]
> Feels like it should be a Signal as well here.

my bad, it is a signal, you just get it by flatmapping with a stream

**Comment by yurique** [comment]
To clarify. Do you get this in your app with literally this?
```scala
      .flatMap { t =>
        EventStream.fromValue("test")
      }
```

Or is it some different stream? Maybe one that also (indirectly) used to construct the `$client` in the first place?

**Comment by pbuszka** [comment]
yes I do, and the result is an unused variable. But before there are a lot of things going on including other flatMaps so it is probably something upstream that is the root cause

**Comment by pbuszka** [comment]
I do some testing to narrow down which part upstream generates the problem. In all simple setups of the upstream Signal there were no problems, so the thing appears when upstream is some corner case. Without a reproducible code it's like looking for a needle in a haystack so I think it's on me to find it as I have the only non-working case :)

**Comment by raquo** [comment]
@pbuszka I've just reviewed the implementation of `SwitchStreamStrategy`, and I do see a bit of an oddity there:

```scala
  private[this] var maybeCurrentEventStream: js.UndefOr[Try[EventStream[O]]] = parent match {
    case signal: Signal[I @unchecked] => signal.tryNow().map(makeStream)
    case _ => js.undefined
  }
```

This would cause the parent signal's `tryNow()` to be called (triggering an evaluation of the parent signal's initial value) before the parent signal was actually started (assuming other code didn't start parent signal already), even if the resulting flattened stream was never "used" (started). I would call it a contract violation, yes. For that to become a problem, the parent signal's initial value would need to depend on the exact point of time when that initial value is evaluated (e.g. if you get that initial value from the DOM somehow).

I'm not sure if this is what's causing your issue, but that's definitely one thing I should fix.

**Comment by pbuszka** [comment]
So I digged deeper and made it a bit simpler and found out the necessary conditions to trigger the problem in my system:
```
case class AdminToken(...)

val $dt: SignalOptionT[AdminToken] = ??? // here still a couple things going on upstream

val i1 = $dt.value
val i2 = $dt.value
// required to BREAK
// val i2 = $dt.value.map(_.map(_ => "a"))

val testResult =
  Val("test1")
    .combineWith(i2)
// required to BREAK
//    .flatMap { t =>
//      EventStream.fromValue("test")
//    }(FlattenStrategy.SwitchStreamStrategy)
// if flatMap strategy changed to ConcurrentStreamStrategy then the problem disappears

//breaks
div(
      child.maybe <-- EventStream
        .fromValue("t")
        .delay(100)
        // .delay(4000) // required TO BREAK
        .toWeakSignal
        .signal
        .split(_ => ())((_, _, s) => {
          div(
            i1 --> { client =>
              error(s"INSIDE test monitoring mapped val 1 ${client}")
            },
            i2 --> { client =>
              error(s"INSIDE test monitoring mapped val 2 ${client}")
            }
          )
        })
)
```
Suming up:
1) upstream a Signal which changes from None to Some after some delay
2) the signal must be mapped 
3) it's observer must be in a div which is delayed as well compared to "touching" of the i2 signal by testResult

I think it roughly fits the potential cause you mentioned above.

**Comment by raquo** [comment]
Do any of the signals involved have an initial value that can't be reduced to a constant? For example, stuff like`stream.startWith(thisNode.ref.clientWidth)`? As opposed to `stream.startWith(1)`? Note that the `startWith` argument is lazily evaluated.

**Comment by pbuszka** [comment]
Nothing which depends on the DOM for sure but effectively there is a chain 
`Var.signal.flatMap(???: EventStream).toWeakSignal.flatMap(???: EventStream).toWeakSignal`
leading to `val $dt: SignalOptionT[AdminToken]` and definitly timing is significant in producing good or bad result.

I am not sure what is just a random correlation and what is an actual cause. The ConcurrentStreamStrategy "fix" kind of confirms your diagnosis as it does not evaluate tryNow() of the parent as SwitchStreamStrategy does and this seems to be the key trigger.

Anyway if this stuff will be changed later then I will stick with the workaround of ConcurrentStreamStrategy for now. I'll try to prepare a test case for laminar which reproduces this problem.

**Comment by pbuszka** [comment]
The part which really left me completly confused is the .map requirement to trigger the problem

**Comment by raquo** [comment]
In the graph of observables events normally propagate depth-first (disregarding special cases like `combineWith`), so depending on `stream` vs `stream.map(...)` can produce different schedules. Again, not something that I would expect to break things in the manner you descrive. I'm still baffled by the whole thing, not sure what could be causing it...

**Comment by raquo** [comment]
We've confirmed that this branch solves the problem: https://github.com/raquo/Airstream/tree/rough-flatten-fix

I'll release this as Airstream 0.13.1 soon.

**Comment by raquo** [comment]
Release ETA in a couple days, got sidetracked with other stuff.

**Comment by pbuszka** [comment]
Any update on ETA ,if you think it's longer then a week please let me know then I'll make my own module from the fork.

**Comment by raquo** [comment]
Hey @pbuszka, sorry for the delay, I got hit by some urgent personal issues.

I should be able to make a release this weekend – it turned out that some other flatten observables also need this fix, so I need a bit more code and more tests for this fix to be comprehensive.

**Comment by pbuszka** [comment]
Cool, if I can do something to help let me know.

**Comment by raquo** [comment]
@pbuszka I've just released Airstream branch `rough-flatten-fix` as `0.13.2-RC1`. It will be up on Maven Central shortly.
